ilastik/applets/neuralNetwork/modelStateControl.py

```diff
@@ -0,0 +1,414 @@
+
+        def _onUploadDone():
+            self.uploadDone.disconnect()
+            dialog.accept()
+
+        self.uploadDone.connect(_onUploadDone)
+
+        def _onDone(fut):
+            self.uploadDone.emit()
+            if fut.cancelled():
+                return
+
+            if fut.exception():
+                self._showErrorMessage(fut.exception())
+
+        modelInfo.add_done_callback(_onDone)
+
+    def uploadModelClicked(self):
+        try:
+            self._uploadModel()
+        except Exception as e:
+            self._showErrorMessage(e)
+
+    def onModelInfoRequested(self):
+        model_uri = self.modelSourceEdit.getModelSource().strip()
+        if not model_uri:
+            # try select file from file chooser
+            model_uri = self.getModelToOpen(self)
+        if not model_uri:
+            return
+
+        model_info = load_raw_resource_description(model_uri)
+
+        self.modelSourceEdit.setModelInfo(model_uri, model_info)
+        # check model is broadly compatible with ilastik
+        try:
+            compatibility_checks = self.checkModelCompatibility(model_info)
+
+            if compatibility_checks:
+                reasons = "\n".join(r["reason"] for r in compatibility_checks)
+                QMessageBox.information(
+                    self,
+                    "Model incompatible",
+                    f"Model incompatible, reasons:\n\n{reasons}.\nPlease select a different model.",
+                )
+                reasons_log = " - ".join(r["reason"] for r in compatibility_checks)
+                logger.debug(f"Incompatible model from {model_uri}. Reasons: {reasons_log}")
+                self.modelSourceEdit.setModelIncompatibleState(model_uri, model_info, reasons)
+                return
+        except Exception as e:
+            self._showErrorMessage(e)
+            return
+
+        try:
+            downloader = self.resolveModel(model_uri, model_info)
+        except Exception as e:
+            self._showErrorMessage(e)
+
+    def resolveModel(self, model_uri, model_info) -> BioImageDownloader:
+        """Initiate model download"""
+        cancelSrc = CancellationTokenSource()
+        downloader = BioImageDownloader(self, model_uri, cancelSrc.token)
+        dialog = PercentProgressDialog(self, title="Downloading model", secondary_bar=True)
+        dialog.rejected.connect(cancelSrc.cancel)
+        dialog.show()
+        downloader.finished.connect(dialog.accept)
+        downloader.error.connect(self._showErrorMessage)
+        downloader.progress0.connect(dialog.updateProgress)
+        downloader.progress1.connect(partial(dialog.updateProgress, bar=BarId.bar1))
+        downloader.currentUri.connect(partial(dialog.updateBarFormat, bar=BarId.bar1))
+        downloader.dataAvailable.connect(partial(self._setAndUploadModel, model_uri, model_info))
+        downloader.start()
+        return downloader
+
+    @staticmethod
+    def _check_model_compatible(model_info):
```

Comment:
Maybe rename to `_required_compatibility_errors()` since it returns a sequence of "error description" objects? The current name implied some "check" action on `model_info` itself. 

Also, `if ...: checks.append(...)` could be replaced with generator:
```python
if cond1:
    yield ...
if cond2:
    yield ...
```

Reply:
I agree the name should be changed - but I'm not sure I'm on board with making it a generator - in the end I'd be interested to see all errors at once. Could you elaborate the advantage of the generator here?

Reply:
I think I'll go with the simpler approach here :)